packages/@storefront/core/src/services/url.ts

```diff
@@ -24,166 +20,96 @@ class UrlService extends BaseService<UrlService.Options> {
-        break;
-      case Routes.PAST_PURCHASE:
-        request = Utils.pastPurchaseStateToRequest(<UrlBeautifier.SearchUrlState>urlState, this.app.flux.store.getState());
-        newState = Utils.mergePastPurchaseState(this.app.flux.store.getState(), <UrlBeautifier.SearchUrlState>urlState);
-        this.refreshState(newState);
-        this.app.flux.store.dispatch(this.app.flux.actions.fetchPastPurchaseProducts({ request }));
-        break;
-      case Routes.DETAILS:
-        this.app.flux.store.dispatch(
-          this.app.flux.actions.fetchProductDetails({ id: (<UrlBeautifier.DetailsUrlState>urlState).data.id })
-        );
-        break;
-    }
+    this.historyListener(this.rewind);
   }
 
-  generateRoutes() {
-    const basePath = Utils.getBasePath();
-    const routes = this.opts.routes;
-
-    return Object.keys(routes).reduce(
-      (generatedRoutes: any, prop) => Object.assign(generatedRoutes, { [prop]: basePath + routes[prop] }),
-      {}
-    );
+  parse = (url: string) => {
+    return this.beautifier.parse<UrlBeautifier.SearchUrlState>(url);
   }
 
-  listenForHistoryChange = () => {
-    this.app.flux.on(Events.HISTORY_SAVE, this.updateHistory);
-    this.app.flux.on(Events.HISTORY_REPLACE, this.buildUrlAndReplaceHistory);
+  build = (route: string, state: Store.State) => {
+    return this.beautifier.build(route, this.urlState[route](state));
   }
 
-  updateHistory = ({ state, route }: { state: Store.State, route: string }) => {
-    const url = this.beautifier.build(route, this.urlState[route](state));
-
+  pushState = () => {
     if (typeof this.opts.urlHandler === 'function') {
-      this.opts.urlHandler(url);
-    } else if (typeof this.opts.redirects === 'function' && this.opts.redirects(url)) {
-      WINDOW().location.assign(this.opts.redirects(url));
-    } else if (this.opts.redirects[url]) {
-      WINDOW().location.assign(this.opts.redirects[url]);
+      return (data, title, url) => this.opts.urlHandler(url);
+    } else if (this.opts.history && typeof this.opts.history.pushState === 'function') {
+      return this.opts.history.pushState;
     } else {
-      try {
-        const oldUrl = WINDOW().location.href;
-        WINDOW().history.pushState(
-          { url, state: this.filterState(this.app.flux.store.getState()), app: STOREFRONT_APP_ID },
-          '',
-          url
-        );
-
-        const newUrl = WINDOW().location.href;
-        this.emitUrlUpdated(oldUrl, newUrl, url);
-        this.handleUrl();
-      } catch (e) {
-        this.app.log.warn('unable to push state to browser history', e);
+      return (data, title, url) => {
+        const redirectFnResult = typeof this.opts.redirects === 'function'
+          ? this.opts.redirects(url)
+          : this.opts.redirects[url];
+        if (redirectFnResult) {
+          return WINDOW().location.assign(redirectFnResult);
+        } else {
+          return WINDOW().history.pushState(data, title, url);
```

Comment:
@Vic-Dev Can we inject the `window.history` object as a default at instantiation, rather than accessing it within each instance method?

For example, the methods which currently access `WINDOW().history.<method>` would be updated to read `this.defaults.history.<method>` (or similar).

Reply:
@jrmykolyn `history` methods have to be called on `window` iirc, so can you be more specific with what you're suggesting?

Reply:
@Vic-Dev I believe that the history methods must be called in the context of the history object itself, or of an object that implements the same interface.

That aside, I should have said: what about something like this:

```
this.defaults.history = <implementation-provided-api> || WINDOW().history;
```

From what I can tell, this should still work.

Reply:
This has been addressed by setting up the `history` API methods in the `UrlService` constructor. Resolving.